{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'nwhy'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-25-b49e8945eb2b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mscipy\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msparse\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mcsr_matrix\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtriu\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mcollections\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mOrderedDict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mnwhy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'nwhy'"
     ]
    }
   ],
   "source": [
    "import hypernetx as hnx\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import zscore\n",
    "from scipy.sparse import csr_matrix, triu\n",
    "from collections import OrderedDict\n",
    "import nwhy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Generating a hypergraph from transcriptomics data\n",
    "In computational biology measuring expression of gene transcripts is called \"transcriptomics\". Given a biological sample it's possible to measure the expression level of genes through....\n",
    "\n",
    "This data file contains log2 fold change values (columns ending in \"\\__b\") as well as significance values (columns ending in \"\\__p\") for each of 279 experimental conditions. Each row corresponds to a gene, with the name of the gene as the first element in the row. \n",
    "\n",
    "We will create a hypergraph from these transcriptomics data where the vertices represent *conditions* and the hyperedges represent *genes*. A condition vertex will be contained within a gene hyperedge if the z-score (condition-wise) of the log2 fold change is $\\geq 2$ and the p-value is $\\leq 0.05$. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_0__b</th>\n",
       "      <th>condition_0__p</th>\n",
       "      <th>condition_1__b</th>\n",
       "      <th>condition_1__p</th>\n",
       "      <th>condition_2__b</th>\n",
       "      <th>condition_2__p</th>\n",
       "      <th>condition_3__b</th>\n",
       "      <th>condition_3__p</th>\n",
       "      <th>condition_4__b</th>\n",
       "      <th>condition_4__p</th>\n",
       "      <th>...</th>\n",
       "      <th>condition_274__b</th>\n",
       "      <th>condition_274__p</th>\n",
       "      <th>condition_275__b</th>\n",
       "      <th>condition_275__p</th>\n",
       "      <th>condition_276__b</th>\n",
       "      <th>condition_276__p</th>\n",
       "      <th>condition_277__b</th>\n",
       "      <th>condition_277__p</th>\n",
       "      <th>condition_278__b</th>\n",
       "      <th>condition_278__p</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>-0.053539</td>\n",
       "      <td>0.897905</td>\n",
       "      <td>-0.021629</td>\n",
       "      <td>0.971758</td>\n",
       "      <td>0.069450</td>\n",
       "      <td>0.893481</td>\n",
       "      <td>0.390366</td>\n",
       "      <td>0.150667</td>\n",
       "      <td>0.073836</td>\n",
       "      <td>0.705193</td>\n",
       "      <td>...</td>\n",
       "      <td>0.818601</td>\n",
       "      <td>0.199452</td>\n",
       "      <td>0.675964</td>\n",
       "      <td>0.000681</td>\n",
       "      <td>1.295524</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>1.474642</td>\n",
       "      <td>0.000588</td>\n",
       "      <td>1.359757</td>\n",
       "      <td>0.012836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.031504</td>\n",
       "      <td>0.896717</td>\n",
       "      <td>0.131252</td>\n",
       "      <td>0.409202</td>\n",
       "      <td>0.309998</td>\n",
       "      <td>0.244921</td>\n",
       "      <td>0.401375</td>\n",
       "      <td>0.037433</td>\n",
       "      <td>0.165346</td>\n",
       "      <td>0.354110</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.549177</td>\n",
       "      <td>0.038906</td>\n",
       "      <td>-0.788553</td>\n",
       "      <td>0.123942</td>\n",
       "      <td>-0.743268</td>\n",
       "      <td>0.038090</td>\n",
       "      <td>-0.632166</td>\n",
       "      <td>0.112165</td>\n",
       "      <td>-1.220868</td>\n",
       "      <td>0.041723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>-0.041660</td>\n",
       "      <td>0.837731</td>\n",
       "      <td>0.031732</td>\n",
       "      <td>0.905520</td>\n",
       "      <td>-0.106712</td>\n",
       "      <td>0.351493</td>\n",
       "      <td>-0.224353</td>\n",
       "      <td>0.169825</td>\n",
       "      <td>-0.045614</td>\n",
       "      <td>0.834415</td>\n",
       "      <td>...</td>\n",
       "      <td>0.036319</td>\n",
       "      <td>0.768323</td>\n",
       "      <td>0.030294</td>\n",
       "      <td>0.781156</td>\n",
       "      <td>0.030260</td>\n",
       "      <td>0.748385</td>\n",
       "      <td>0.020523</td>\n",
       "      <td>0.891356</td>\n",
       "      <td>0.009379</td>\n",
       "      <td>0.963882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>0.139425</td>\n",
       "      <td>0.507731</td>\n",
       "      <td>0.148185</td>\n",
       "      <td>0.805341</td>\n",
       "      <td>0.251240</td>\n",
       "      <td>0.196918</td>\n",
       "      <td>0.290396</td>\n",
       "      <td>0.058642</td>\n",
       "      <td>0.383917</td>\n",
       "      <td>0.082190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.235774</td>\n",
       "      <td>0.359214</td>\n",
       "      <td>0.871901</td>\n",
       "      <td>0.009966</td>\n",
       "      <td>1.341327</td>\n",
       "      <td>0.034060</td>\n",
       "      <td>-0.485618</td>\n",
       "      <td>0.168506</td>\n",
       "      <td>-1.095028</td>\n",
       "      <td>0.100790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>0.139837</td>\n",
       "      <td>0.470499</td>\n",
       "      <td>0.005684</td>\n",
       "      <td>0.988458</td>\n",
       "      <td>0.077773</td>\n",
       "      <td>0.496487</td>\n",
       "      <td>0.128518</td>\n",
       "      <td>0.292702</td>\n",
       "      <td>0.132485</td>\n",
       "      <td>0.179721</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.181211</td>\n",
       "      <td>0.237053</td>\n",
       "      <td>0.389658</td>\n",
       "      <td>0.346917</td>\n",
       "      <td>-0.150656</td>\n",
       "      <td>0.560934</td>\n",
       "      <td>-0.244075</td>\n",
       "      <td>0.553876</td>\n",
       "      <td>-0.914955</td>\n",
       "      <td>0.221456</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 558 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       condition_0__b  condition_0__p  condition_1__b  condition_1__p  \\\n",
       "AAAS        -0.053539        0.897905       -0.021629        0.971758   \n",
       "AACS         0.031504        0.896717        0.131252        0.409202   \n",
       "AADAC       -0.041660        0.837731        0.031732        0.905520   \n",
       "AAK1         0.139425        0.507731        0.148185        0.805341   \n",
       "AAMP         0.139837        0.470499        0.005684        0.988458   \n",
       "\n",
       "       condition_2__b  condition_2__p  condition_3__b  condition_3__p  \\\n",
       "AAAS         0.069450        0.893481        0.390366        0.150667   \n",
       "AACS         0.309998        0.244921        0.401375        0.037433   \n",
       "AADAC       -0.106712        0.351493       -0.224353        0.169825   \n",
       "AAK1         0.251240        0.196918        0.290396        0.058642   \n",
       "AAMP         0.077773        0.496487        0.128518        0.292702   \n",
       "\n",
       "       condition_4__b  condition_4__p  ...  condition_274__b  \\\n",
       "AAAS         0.073836        0.705193  ...          0.818601   \n",
       "AACS         0.165346        0.354110  ...         -1.549177   \n",
       "AADAC       -0.045614        0.834415  ...          0.036319   \n",
       "AAK1         0.383917        0.082190  ...          0.235774   \n",
       "AAMP         0.132485        0.179721  ...         -1.181211   \n",
       "\n",
       "       condition_274__p  condition_275__b  condition_275__p  condition_276__b  \\\n",
       "AAAS           0.199452          0.675964          0.000681          1.295524   \n",
       "AACS           0.038906         -0.788553          0.123942         -0.743268   \n",
       "AADAC          0.768323          0.030294          0.781156          0.030260   \n",
       "AAK1           0.359214          0.871901          0.009966          1.341327   \n",
       "AAMP           0.237053          0.389658          0.346917         -0.150656   \n",
       "\n",
       "       condition_276__p  condition_277__b  condition_277__p  condition_278__b  \\\n",
       "AAAS           0.000307          1.474642          0.000588          1.359757   \n",
       "AACS           0.038090         -0.632166          0.112165         -1.220868   \n",
       "AADAC          0.748385          0.020523          0.891356          0.009379   \n",
       "AAK1           0.034060         -0.485618          0.168506         -1.095028   \n",
       "AAMP           0.560934         -0.244075          0.553876         -0.914955   \n",
       "\n",
       "       condition_278__p  \n",
       "AAAS           0.012836  \n",
       "AACS           0.041723  \n",
       "AADAC          0.963882  \n",
       "AAK1           0.100790  \n",
       "AAMP           0.221456  \n",
       "\n",
       "[5 rows x 558 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trans_datafile ='bigTrans3_anon.csv'\n",
    "\n",
    "## Read the data file into a pandas data frame\n",
    "tdf = pd.read_csv(f'{trans_datafile}', sep=',', index_col=0)\n",
    "\n",
    "tdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to construct the hypergraph using the `from_dataframe` method in HyperNetX we need to separate the columns and create one dataframe for the log2 fold changes and one for the p-values. We will also need a translation from the p-value column name to the corresponding log2 fold change column name. This will be the `p_to_b_dict` dictionary."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "b_cols = []\n",
    "p_cols = []\n",
    "\n",
    "p_to_b_dict = {}\n",
    "\n",
    "for col in tdf.columns:\n",
    "    if col[-1] == 'b':\n",
    "        b_cols += [col]\n",
    "    elif col[-1] == 'p':\n",
    "        p_cols += [col]\n",
    "    else:\n",
    "        print(col) #shouldn't happen\n",
    "        \n",
    "## In order to apply a filter based on p-values to the log2 fold change values later we create `tdfp_renamed` \n",
    "## which has the column names of the p-value dataframe align with the column names of `tdfb`.\n",
    "tdfb = tdf[b_cols].rename(columns=lambda x : x.replace('__b','')) #log2 fold change data frame\n",
    "tdfp = tdf[p_cols].rename(columns=lambda x : x.replace('__p','')) #p-value data frame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_0</th>\n",
       "      <th>condition_1</th>\n",
       "      <th>condition_2</th>\n",
       "      <th>condition_3</th>\n",
       "      <th>condition_4</th>\n",
       "      <th>condition_5</th>\n",
       "      <th>condition_6</th>\n",
       "      <th>condition_7</th>\n",
       "      <th>condition_8</th>\n",
       "      <th>condition_9</th>\n",
       "      <th>...</th>\n",
       "      <th>condition_269</th>\n",
       "      <th>condition_270</th>\n",
       "      <th>condition_271</th>\n",
       "      <th>condition_272</th>\n",
       "      <th>condition_273</th>\n",
       "      <th>condition_274</th>\n",
       "      <th>condition_275</th>\n",
       "      <th>condition_276</th>\n",
       "      <th>condition_277</th>\n",
       "      <th>condition_278</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>-0.053539</td>\n",
       "      <td>-0.021629</td>\n",
       "      <td>0.069450</td>\n",
       "      <td>0.390366</td>\n",
       "      <td>0.073836</td>\n",
       "      <td>-0.442246</td>\n",
       "      <td>-0.082826</td>\n",
       "      <td>-0.422886</td>\n",
       "      <td>-0.045865</td>\n",
       "      <td>0.308248</td>\n",
       "      <td>...</td>\n",
       "      <td>0.110832</td>\n",
       "      <td>0.106307</td>\n",
       "      <td>-0.164211</td>\n",
       "      <td>0.384582</td>\n",
       "      <td>0.661209</td>\n",
       "      <td>0.818601</td>\n",
       "      <td>0.675964</td>\n",
       "      <td>1.295524</td>\n",
       "      <td>1.474642</td>\n",
       "      <td>1.359757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.031504</td>\n",
       "      <td>0.131252</td>\n",
       "      <td>0.309998</td>\n",
       "      <td>0.401375</td>\n",
       "      <td>0.165346</td>\n",
       "      <td>0.079663</td>\n",
       "      <td>-0.291830</td>\n",
       "      <td>-0.386167</td>\n",
       "      <td>0.207874</td>\n",
       "      <td>0.335756</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.786648</td>\n",
       "      <td>-0.234909</td>\n",
       "      <td>-0.942312</td>\n",
       "      <td>-0.758240</td>\n",
       "      <td>-0.742127</td>\n",
       "      <td>-1.549177</td>\n",
       "      <td>-0.788553</td>\n",
       "      <td>-0.743268</td>\n",
       "      <td>-0.632166</td>\n",
       "      <td>-1.220868</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>-0.041660</td>\n",
       "      <td>0.031732</td>\n",
       "      <td>-0.106712</td>\n",
       "      <td>-0.224353</td>\n",
       "      <td>-0.045614</td>\n",
       "      <td>-0.890692</td>\n",
       "      <td>-0.040650</td>\n",
       "      <td>0.053375</td>\n",
       "      <td>-0.220946</td>\n",
       "      <td>-0.279067</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.030368</td>\n",
       "      <td>-0.041964</td>\n",
       "      <td>0.070461</td>\n",
       "      <td>0.062098</td>\n",
       "      <td>0.027516</td>\n",
       "      <td>0.036319</td>\n",
       "      <td>0.030294</td>\n",
       "      <td>0.030260</td>\n",
       "      <td>0.020523</td>\n",
       "      <td>0.009379</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>0.139425</td>\n",
       "      <td>0.148185</td>\n",
       "      <td>0.251240</td>\n",
       "      <td>0.290396</td>\n",
       "      <td>0.383917</td>\n",
       "      <td>0.680079</td>\n",
       "      <td>-0.185509</td>\n",
       "      <td>0.495312</td>\n",
       "      <td>-0.189335</td>\n",
       "      <td>0.234855</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.645084</td>\n",
       "      <td>-0.540221</td>\n",
       "      <td>0.257469</td>\n",
       "      <td>-0.243665</td>\n",
       "      <td>-0.326640</td>\n",
       "      <td>0.235774</td>\n",
       "      <td>0.871901</td>\n",
       "      <td>1.341327</td>\n",
       "      <td>-0.485618</td>\n",
       "      <td>-1.095028</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>0.139837</td>\n",
       "      <td>0.005684</td>\n",
       "      <td>0.077773</td>\n",
       "      <td>0.128518</td>\n",
       "      <td>0.132485</td>\n",
       "      <td>0.122469</td>\n",
       "      <td>0.025393</td>\n",
       "      <td>0.055635</td>\n",
       "      <td>0.113951</td>\n",
       "      <td>0.080511</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.147891</td>\n",
       "      <td>-0.329626</td>\n",
       "      <td>-0.670756</td>\n",
       "      <td>-0.411586</td>\n",
       "      <td>-0.751645</td>\n",
       "      <td>-1.181211</td>\n",
       "      <td>0.389658</td>\n",
       "      <td>-0.150656</td>\n",
       "      <td>-0.244075</td>\n",
       "      <td>-0.914955</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 279 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       condition_0  condition_1  condition_2  condition_3  condition_4  \\\n",
       "AAAS     -0.053539    -0.021629     0.069450     0.390366     0.073836   \n",
       "AACS      0.031504     0.131252     0.309998     0.401375     0.165346   \n",
       "AADAC    -0.041660     0.031732    -0.106712    -0.224353    -0.045614   \n",
       "AAK1      0.139425     0.148185     0.251240     0.290396     0.383917   \n",
       "AAMP      0.139837     0.005684     0.077773     0.128518     0.132485   \n",
       "\n",
       "       condition_5  condition_6  condition_7  condition_8  condition_9  ...  \\\n",
       "AAAS     -0.442246    -0.082826    -0.422886    -0.045865     0.308248  ...   \n",
       "AACS      0.079663    -0.291830    -0.386167     0.207874     0.335756  ...   \n",
       "AADAC    -0.890692    -0.040650     0.053375    -0.220946    -0.279067  ...   \n",
       "AAK1      0.680079    -0.185509     0.495312    -0.189335     0.234855  ...   \n",
       "AAMP      0.122469     0.025393     0.055635     0.113951     0.080511  ...   \n",
       "\n",
       "       condition_269  condition_270  condition_271  condition_272  \\\n",
       "AAAS        0.110832       0.106307      -0.164211       0.384582   \n",
       "AACS       -0.786648      -0.234909      -0.942312      -0.758240   \n",
       "AADAC      -0.030368      -0.041964       0.070461       0.062098   \n",
       "AAK1       -1.645084      -0.540221       0.257469      -0.243665   \n",
       "AAMP       -1.147891      -0.329626      -0.670756      -0.411586   \n",
       "\n",
       "       condition_273  condition_274  condition_275  condition_276  \\\n",
       "AAAS        0.661209       0.818601       0.675964       1.295524   \n",
       "AACS       -0.742127      -1.549177      -0.788553      -0.743268   \n",
       "AADAC       0.027516       0.036319       0.030294       0.030260   \n",
       "AAK1       -0.326640       0.235774       0.871901       1.341327   \n",
       "AAMP       -0.751645      -1.181211       0.389658      -0.150656   \n",
       "\n",
       "       condition_277  condition_278  \n",
       "AAAS        1.474642       1.359757  \n",
       "AACS       -0.632166      -1.220868  \n",
       "AADAC       0.020523       0.009379  \n",
       "AAK1       -0.485618      -1.095028  \n",
       "AAMP       -0.244075      -0.914955  \n",
       "\n",
       "[5 rows x 279 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdfb.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_0</th>\n",
       "      <th>condition_1</th>\n",
       "      <th>condition_2</th>\n",
       "      <th>condition_3</th>\n",
       "      <th>condition_4</th>\n",
       "      <th>condition_5</th>\n",
       "      <th>condition_6</th>\n",
       "      <th>condition_7</th>\n",
       "      <th>condition_8</th>\n",
       "      <th>condition_9</th>\n",
       "      <th>...</th>\n",
       "      <th>condition_269</th>\n",
       "      <th>condition_270</th>\n",
       "      <th>condition_271</th>\n",
       "      <th>condition_272</th>\n",
       "      <th>condition_273</th>\n",
       "      <th>condition_274</th>\n",
       "      <th>condition_275</th>\n",
       "      <th>condition_276</th>\n",
       "      <th>condition_277</th>\n",
       "      <th>condition_278</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>0.897905</td>\n",
       "      <td>0.971758</td>\n",
       "      <td>0.893481</td>\n",
       "      <td>0.150667</td>\n",
       "      <td>0.705193</td>\n",
       "      <td>0.020239</td>\n",
       "      <td>0.814719</td>\n",
       "      <td>0.026202</td>\n",
       "      <td>0.919330</td>\n",
       "      <td>0.120621</td>\n",
       "      <td>...</td>\n",
       "      <td>0.377591</td>\n",
       "      <td>0.463682</td>\n",
       "      <td>0.594456</td>\n",
       "      <td>0.002115</td>\n",
       "      <td>0.046480</td>\n",
       "      <td>0.199452</td>\n",
       "      <td>0.000681</td>\n",
       "      <td>0.000307</td>\n",
       "      <td>0.000588</td>\n",
       "      <td>0.012836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.896717</td>\n",
       "      <td>0.409202</td>\n",
       "      <td>0.244921</td>\n",
       "      <td>0.037433</td>\n",
       "      <td>0.354110</td>\n",
       "      <td>0.499078</td>\n",
       "      <td>0.189763</td>\n",
       "      <td>0.015905</td>\n",
       "      <td>0.191275</td>\n",
       "      <td>0.040680</td>\n",
       "      <td>...</td>\n",
       "      <td>0.009926</td>\n",
       "      <td>0.194330</td>\n",
       "      <td>0.068695</td>\n",
       "      <td>0.000293</td>\n",
       "      <td>0.013817</td>\n",
       "      <td>0.038906</td>\n",
       "      <td>0.123942</td>\n",
       "      <td>0.038090</td>\n",
       "      <td>0.112165</td>\n",
       "      <td>0.041723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>0.837731</td>\n",
       "      <td>0.905520</td>\n",
       "      <td>0.351493</td>\n",
       "      <td>0.169825</td>\n",
       "      <td>0.834415</td>\n",
       "      <td>0.000132</td>\n",
       "      <td>0.794360</td>\n",
       "      <td>0.563794</td>\n",
       "      <td>0.119895</td>\n",
       "      <td>0.025462</td>\n",
       "      <td>...</td>\n",
       "      <td>0.602196</td>\n",
       "      <td>0.647917</td>\n",
       "      <td>0.483492</td>\n",
       "      <td>0.498501</td>\n",
       "      <td>0.772614</td>\n",
       "      <td>0.768323</td>\n",
       "      <td>0.781156</td>\n",
       "      <td>0.748385</td>\n",
       "      <td>0.891356</td>\n",
       "      <td>0.963882</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>0.507731</td>\n",
       "      <td>0.805341</td>\n",
       "      <td>0.196918</td>\n",
       "      <td>0.058642</td>\n",
       "      <td>0.082190</td>\n",
       "      <td>0.000294</td>\n",
       "      <td>0.383349</td>\n",
       "      <td>0.009896</td>\n",
       "      <td>0.147586</td>\n",
       "      <td>0.011289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007530</td>\n",
       "      <td>0.219397</td>\n",
       "      <td>0.304758</td>\n",
       "      <td>0.006622</td>\n",
       "      <td>0.057828</td>\n",
       "      <td>0.359214</td>\n",
       "      <td>0.009966</td>\n",
       "      <td>0.034060</td>\n",
       "      <td>0.168506</td>\n",
       "      <td>0.100790</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>0.470499</td>\n",
       "      <td>0.988458</td>\n",
       "      <td>0.496487</td>\n",
       "      <td>0.292702</td>\n",
       "      <td>0.179721</td>\n",
       "      <td>0.264728</td>\n",
       "      <td>0.849092</td>\n",
       "      <td>0.624992</td>\n",
       "      <td>0.311908</td>\n",
       "      <td>0.220028</td>\n",
       "      <td>...</td>\n",
       "      <td>0.023810</td>\n",
       "      <td>0.519994</td>\n",
       "      <td>0.081114</td>\n",
       "      <td>0.013810</td>\n",
       "      <td>0.007500</td>\n",
       "      <td>0.237053</td>\n",
       "      <td>0.346917</td>\n",
       "      <td>0.560934</td>\n",
       "      <td>0.553876</td>\n",
       "      <td>0.221456</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 279 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       condition_0  condition_1  condition_2  condition_3  condition_4  \\\n",
       "AAAS      0.897905     0.971758     0.893481     0.150667     0.705193   \n",
       "AACS      0.896717     0.409202     0.244921     0.037433     0.354110   \n",
       "AADAC     0.837731     0.905520     0.351493     0.169825     0.834415   \n",
       "AAK1      0.507731     0.805341     0.196918     0.058642     0.082190   \n",
       "AAMP      0.470499     0.988458     0.496487     0.292702     0.179721   \n",
       "\n",
       "       condition_5  condition_6  condition_7  condition_8  condition_9  ...  \\\n",
       "AAAS      0.020239     0.814719     0.026202     0.919330     0.120621  ...   \n",
       "AACS      0.499078     0.189763     0.015905     0.191275     0.040680  ...   \n",
       "AADAC     0.000132     0.794360     0.563794     0.119895     0.025462  ...   \n",
       "AAK1      0.000294     0.383349     0.009896     0.147586     0.011289  ...   \n",
       "AAMP      0.264728     0.849092     0.624992     0.311908     0.220028  ...   \n",
       "\n",
       "       condition_269  condition_270  condition_271  condition_272  \\\n",
       "AAAS        0.377591       0.463682       0.594456       0.002115   \n",
       "AACS        0.009926       0.194330       0.068695       0.000293   \n",
       "AADAC       0.602196       0.647917       0.483492       0.498501   \n",
       "AAK1        0.007530       0.219397       0.304758       0.006622   \n",
       "AAMP        0.023810       0.519994       0.081114       0.013810   \n",
       "\n",
       "       condition_273  condition_274  condition_275  condition_276  \\\n",
       "AAAS        0.046480       0.199452       0.000681       0.000307   \n",
       "AACS        0.013817       0.038906       0.123942       0.038090   \n",
       "AADAC       0.772614       0.768323       0.781156       0.748385   \n",
       "AAK1        0.057828       0.359214       0.009966       0.034060   \n",
       "AAMP        0.007500       0.237053       0.346917       0.560934   \n",
       "\n",
       "       condition_277  condition_278  \n",
       "AAAS        0.000588       0.012836  \n",
       "AACS        0.112165       0.041723  \n",
       "AADAC       0.891356       0.963882  \n",
       "AAK1        0.168506       0.100790  \n",
       "AAMP        0.553876       0.221456  \n",
       "\n",
       "[5 rows x 279 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdfp.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We check to see if there are any NA values in `tdfb` (there are). If so, we must fill the NAs in order to compute z-scores. We fill the NAs with the mean of each column. Later, when we create the hypergraph, it will be clear why this is a reasonable choice."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdfb.isna().sum().sum()  ## number of Nans\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "tdfb = tdfb.astype(float) # to ensure everything is a float\n",
    "tdfb = tdfb.fillna(tdfb.mean()) # fill the NAs with the mean of each column"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We apply the z-score column-wise and take the absolute value of each resulting entry. It is worth pointing out that filling the NAs with the mean of each column does have an impact on the z-scores (by reducing the standard deviation of the column). However, in this case there are so few NA values and for each affected column there is only one NA, that that we assume the impact is negligible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition_0</th>\n",
       "      <th>condition_1</th>\n",
       "      <th>condition_2</th>\n",
       "      <th>condition_3</th>\n",
       "      <th>condition_4</th>\n",
       "      <th>condition_5</th>\n",
       "      <th>condition_6</th>\n",
       "      <th>condition_7</th>\n",
       "      <th>condition_8</th>\n",
       "      <th>condition_9</th>\n",
       "      <th>...</th>\n",
       "      <th>condition_269</th>\n",
       "      <th>condition_270</th>\n",
       "      <th>condition_271</th>\n",
       "      <th>condition_272</th>\n",
       "      <th>condition_273</th>\n",
       "      <th>condition_274</th>\n",
       "      <th>condition_275</th>\n",
       "      <th>condition_276</th>\n",
       "      <th>condition_277</th>\n",
       "      <th>condition_278</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>0.385565</td>\n",
       "      <td>0.238070</td>\n",
       "      <td>0.443872</td>\n",
       "      <td>1.463033</td>\n",
       "      <td>0.158425</td>\n",
       "      <td>1.145265</td>\n",
       "      <td>0.260249</td>\n",
       "      <td>2.064284</td>\n",
       "      <td>0.379907</td>\n",
       "      <td>0.953880</td>\n",
       "      <td>...</td>\n",
       "      <td>0.385399</td>\n",
       "      <td>0.447249</td>\n",
       "      <td>0.119840</td>\n",
       "      <td>0.836579</td>\n",
       "      <td>1.291596</td>\n",
       "      <td>1.311356</td>\n",
       "      <td>0.969919</td>\n",
       "      <td>1.879485</td>\n",
       "      <td>2.509792</td>\n",
       "      <td>1.883533</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACS</th>\n",
       "      <td>0.164414</td>\n",
       "      <td>1.005465</td>\n",
       "      <td>2.345166</td>\n",
       "      <td>1.507704</td>\n",
       "      <td>0.606929</td>\n",
       "      <td>0.163841</td>\n",
       "      <td>1.217533</td>\n",
       "      <td>1.890122</td>\n",
       "      <td>1.473085</td>\n",
       "      <td>1.046324</td>\n",
       "      <td>...</td>\n",
       "      <td>0.706847</td>\n",
       "      <td>0.358692</td>\n",
       "      <td>1.394904</td>\n",
       "      <td>0.985461</td>\n",
       "      <td>1.094888</td>\n",
       "      <td>1.613561</td>\n",
       "      <td>1.095892</td>\n",
       "      <td>0.670392</td>\n",
       "      <td>0.860191</td>\n",
       "      <td>1.092556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AADAC</th>\n",
       "      <td>0.308742</td>\n",
       "      <td>0.195965</td>\n",
       "      <td>0.948517</td>\n",
       "      <td>1.031329</td>\n",
       "      <td>0.427018</td>\n",
       "      <td>2.270104</td>\n",
       "      <td>0.067074</td>\n",
       "      <td>0.194677</td>\n",
       "      <td>1.658488</td>\n",
       "      <td>1.019828</td>\n",
       "      <td>...</td>\n",
       "      <td>0.213557</td>\n",
       "      <td>0.097038</td>\n",
       "      <td>0.264714</td>\n",
       "      <td>0.322433</td>\n",
       "      <td>0.213951</td>\n",
       "      <td>0.345003</td>\n",
       "      <td>0.059153</td>\n",
       "      <td>0.297044</td>\n",
       "      <td>0.183829</td>\n",
       "      <td>0.326218</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAK1</th>\n",
       "      <td>0.862342</td>\n",
       "      <td>1.143194</td>\n",
       "      <td>1.880740</td>\n",
       "      <td>1.057383</td>\n",
       "      <td>1.678182</td>\n",
       "      <td>1.669868</td>\n",
       "      <td>0.730559</td>\n",
       "      <td>2.290836</td>\n",
       "      <td>1.427638</td>\n",
       "      <td>0.707239</td>\n",
       "      <td>...</td>\n",
       "      <td>1.751575</td>\n",
       "      <td>1.079826</td>\n",
       "      <td>0.571162</td>\n",
       "      <td>0.165057</td>\n",
       "      <td>0.388318</td>\n",
       "      <td>0.591389</td>\n",
       "      <td>1.246303</td>\n",
       "      <td>1.936769</td>\n",
       "      <td>0.625779</td>\n",
       "      <td>0.947431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAMP</th>\n",
       "      <td>0.865006</td>\n",
       "      <td>0.015907</td>\n",
       "      <td>0.509653</td>\n",
       "      <td>0.400525</td>\n",
       "      <td>0.445872</td>\n",
       "      <td>0.271212</td>\n",
       "      <td>0.235416</td>\n",
       "      <td>0.205397</td>\n",
       "      <td>0.787188</td>\n",
       "      <td>0.188556</td>\n",
       "      <td>...</td>\n",
       "      <td>1.146484</td>\n",
       "      <td>0.582409</td>\n",
       "      <td>0.949909</td>\n",
       "      <td>0.432778</td>\n",
       "      <td>1.111073</td>\n",
       "      <td>1.159014</td>\n",
       "      <td>0.566062</td>\n",
       "      <td>0.070777</td>\n",
       "      <td>0.239414</td>\n",
       "      <td>0.739763</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 279 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       condition_0  condition_1  condition_2  condition_3  condition_4  \\\n",
       "AAAS      0.385565     0.238070     0.443872     1.463033     0.158425   \n",
       "AACS      0.164414     1.005465     2.345166     1.507704     0.606929   \n",
       "AADAC     0.308742     0.195965     0.948517     1.031329     0.427018   \n",
       "AAK1      0.862342     1.143194     1.880740     1.057383     1.678182   \n",
       "AAMP      0.865006     0.015907     0.509653     0.400525     0.445872   \n",
       "\n",
       "       condition_5  condition_6  condition_7  condition_8  condition_9  ...  \\\n",
       "AAAS      1.145265     0.260249     2.064284     0.379907     0.953880  ...   \n",
       "AACS      0.163841     1.217533     1.890122     1.473085     1.046324  ...   \n",
       "AADAC     2.270104     0.067074     0.194677     1.658488     1.019828  ...   \n",
       "AAK1      1.669868     0.730559     2.290836     1.427638     0.707239  ...   \n",
       "AAMP      0.271212     0.235416     0.205397     0.787188     0.188556  ...   \n",
       "\n",
       "       condition_269  condition_270  condition_271  condition_272  \\\n",
       "AAAS        0.385399       0.447249       0.119840       0.836579   \n",
       "AACS        0.706847       0.358692       1.394904       0.985461   \n",
       "AADAC       0.213557       0.097038       0.264714       0.322433   \n",
       "AAK1        1.751575       1.079826       0.571162       0.165057   \n",
       "AAMP        1.146484       0.582409       0.949909       0.432778   \n",
       "\n",
       "       condition_273  condition_274  condition_275  condition_276  \\\n",
       "AAAS        1.291596       1.311356       0.969919       1.879485   \n",
       "AACS        1.094888       1.613561       1.095892       0.670392   \n",
       "AADAC       0.213951       0.345003       0.059153       0.297044   \n",
       "AAK1        0.388318       0.591389       1.246303       1.936769   \n",
       "AAMP        1.111073       1.159014       0.566062       0.070777   \n",
       "\n",
       "       condition_277  condition_278  \n",
       "AAAS        2.509792       1.883533  \n",
       "AACS        0.860191       1.092556  \n",
       "AADAC       0.183829       0.326218  \n",
       "AAK1        0.625779       0.947431  \n",
       "AAMP        0.239414       0.739763  \n",
       "\n",
       "[5 rows x 279 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdfbz = abs(tdfb.apply(zscore))\n",
    "tdfbz.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next we filter the z-score dataframe to zero out all of the values corresponding to p-values greater than 0.05. This will allow us to do the z-score > 2 AND p-value < 0.05 criteria for edge creation. `tdfp < 0.05` is a dataframe of `True` and `False` values, based on the filter. Taking `tdfz[tdfp < 0.05]` applies this p-value filter to the z-scores. Any place where there was a `False` will become `NA`. Then fill the NAs with 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AADAC</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>AARS</th>\n",
       "      <th>AASDHPPT</th>\n",
       "      <th>AASS</th>\n",
       "      <th>AATF</th>\n",
       "      <th>AATK</th>\n",
       "      <th>ABAT</th>\n",
       "      <th>...</th>\n",
       "      <th>ZSWIM5</th>\n",
       "      <th>ZSWIM6</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>condition_3</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.13085</td>\n",
       "      <td>2.044925</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.551110</td>\n",
       "      <td>9.441469</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>condition_4</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.988752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>condition_5</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.270104</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>11.733724</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>condition_6</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.403344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>condition_7</th>\n",
       "      <td>2.064284</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.290836</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.671948</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.636012</td>\n",
       "      <td>3.494152</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 8030 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 AAAS     AADAC      AAK1  AAMP     AARS  AASDHPPT  AASS  \\\n",
       "condition_3  0.000000  0.000000  0.000000   0.0  2.13085  2.044925   0.0   \n",
       "condition_4  0.000000  0.000000  0.000000   0.0  0.00000  0.000000   0.0   \n",
       "condition_5  0.000000  2.270104  0.000000   0.0  0.00000  0.000000   0.0   \n",
       "condition_6  0.000000  0.000000  0.000000   0.0  0.00000  0.000000   0.0   \n",
       "condition_7  2.064284  0.000000  2.290836   0.0  0.00000  0.000000   0.0   \n",
       "\n",
       "             AATF      AATK  ABAT  ...  ZSWIM5  ZSWIM6  ZW10  ZWILCH  ZWINT  \\\n",
       "condition_3   0.0  0.000000   0.0  ...     0.0     0.0   0.0     0.0    0.0   \n",
       "condition_4   0.0  0.000000   0.0  ...     0.0     0.0   0.0     0.0    0.0   \n",
       "condition_5   0.0  0.000000   0.0  ...     0.0     0.0   0.0     0.0    0.0   \n",
       "condition_6   0.0  0.000000   0.0  ...     0.0     0.0   0.0     0.0    0.0   \n",
       "condition_7   0.0  2.671948   0.0  ...     0.0     0.0   0.0     0.0    0.0   \n",
       "\n",
       "             ZXDA  ZXDC  ZYG11B       ZYX       ZZZ3  \n",
       "condition_3   0.0   0.0     0.0  2.551110   9.441469  \n",
       "condition_4   0.0   0.0     0.0  0.000000  12.988752  \n",
       "condition_5   0.0   0.0     0.0  0.000000  11.733724  \n",
       "condition_6   0.0   0.0     0.0  0.000000   2.403344  \n",
       "condition_7   0.0   0.0     0.0  3.636012   3.494152  \n",
       "\n",
       "[5 rows x 8030 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdfbz_filtered = tdfbz[tdfp < 0.05]\n",
    "tdfbz_filtered = tdfbz_filtered.fillna(0)\n",
    "tdfbz_filtered = tdfbz_filtered [tdfbz_filtered >= 2.0].fillna(0)\n",
    "tdfbz_filtered = tdfbz_filtered[[c for c in tdfbz.columns if tdfbz_filtered[c].any()]]\n",
    "tdfT = tdfbz_filtered.T\n",
    "tdfbz_filtered = tdfT[[c for c in tdfT.columns if tdfT[c].any()]]\n",
    "tdfbz_filtered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(241, 8030)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tdfbz_filtered.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "labels = OrderedDict([('genes', np.array(tdfbz_filtered.columns)),('conditions', np.array(tdfbz_filtered.index))])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8030, 241)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "arr = tdfbz_filtered.values.transpose();arr.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "E = hnx.StaticEntitySet(arr=arr,labels=labels)\n",
    "H_trans = hnx.Hypergraph(E)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally we are ready to create the hypergraph! We use the `from_dataframe` command with inpt `tdfbz_filtered`, `transpose=True`, and lower threshold of 2.0. We don't apply any transformations because we've already done the necessary filtering and transforms. \n",
    "\n",
    "Now we see why filling the NAs with the mean of each column is reasonable. The z-score of the mean is 0.0 and so any gene that originally had an NA value for a condition will not pass the z-score lower threshold of 2.0 and thus will not appear in the hyperedge corresponding to that condition.\n",
    "\n",
    "In the resulting hypergraph the vertices represent *conditions* and hyperedges represent *genes*. A gene (hyperedge) contains a condition (vertex) if the z-score of the log2 fold change of the gene, within the condition, is greater than or equal to 2.0 __and__ the p-value of the log2 fold change is less than or equal to 0.05. We say that the gene is \"significantly perturbed\" in the condition. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%timeit\n",
    "# 1min 51s ± 469 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)\n",
    "#H_trans = hnx.Hypergraph.from_dataframe(tdfbz_filtered, transpose=True, key = (lambda x : x >= 2.0))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sometimes the dual hypergraph is more useful to consider. In the dual we swap the role of vertices and edges so that the vertices represent genes and the hyperedges represent conditions. The same containment criteria applies, a gene vertex is contained in a condition hyperedge if the gene is significantly perturbed in the condition. We create `H_trans_d` by using the HyperNetX `dual` function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "H_trans_d = H_trans.dual()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Applying some simple hypergraph methods\n",
    "In order to get an understanding of the structure of these hypergraphs we apply some hypernetwork science methods available in HyperNetX. We will apply our initial analysis to the `H_trans` hypergraph and then proceed with some more analysis on both hypergraphs. \n",
    "\n",
    "### Vertex, edge count\n",
    "The `shape` property tells us how many vertices (conditions) and hyperedges (genes) are in the hypergraph. The output is of the form `(num vertices, num hyperedges)`. Notice that there are fewer conditions and fewer genes than in the original `tdfb` data frame. This is becasue some conditions have no genes with z-score $\\geq 2$ and some genes have no conditions in which they are significantly perturbed. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(241, 8030)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "H_trans.shape ## number of genes x number of conditions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `info` function also gives us vertex and edge count (as rows and columns, respectively). In addition we see the aspect ratio, number of nonempty cells in the incidence matrix, and density of the incidence matrix. From the high aspect ratio we see that this is a very \"rectangular\" incidence matrix (aspect ratio of 1.0 is square), and is also quite sparse."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Rows: 241\n",
      "Number of Columns: 8030\n",
      "Aspect Ratio: 0.030012453300124534\n",
      "Number of non-empty Cells: 81147\n",
      "Density: 0.04193145000852612\n"
     ]
    }
   ],
   "source": [
    "print(hnx.info(H_trans))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Degree and edge size distributions\n",
    "The degree of a vertex in `H_trans` is the number of significantly perturbed genes for a given condition. Here we show the distribution of the vertex degrees. It turns out that a histogram visualization is better to see the shape of this distribution."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The size of an edge counts the number of conditions in which a gene is significantly perturbed.  Here we show the distribtion of the degrees on a log-log plot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "M = H_trans.incidence_matrix()\n",
    "degree_distr = np.array(M.sum(axis=1).T)[0]\n",
    "edge_size_distr = np.array(M.sum(axis=0))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 0, '#conditions')"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAv8AAAEWCAYAAADxWPj1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjMsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+AADFEAAAgAElEQVR4nO3debhkZXnv/e9PGkFBBaRVZLAdcE5E0goGBwKoCL5icvSIxyAoCZpg1MREwRhnI2SQmGMcUBEckaAoRzSICA5RwQYBmRTUVloQGrtB1DiA9/vHejZUV1fV3r337j10fT/Xta9d61nTvabnuWtNlapCkiRJ0qbvTvMdgCRJkqS5YfIvSZIkjQmTf0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmFiwyX+SE5O8eb7j2FBJDkvy1Z7unyd5wCxN+9VJ3tc+L0tSSZbM0rR3abFuNhvT06Ylycok+813HBtL//GU5HNJDh0x/LuT/MPcRahhkuydZNUcz3Ojb/8kr0/y4fZ5Vuvn3vhne/0leUKS78zW9BaT2W6XpzH/vZJc1faVZ85TDOvsT0kuS7L3iOFH1rXaOOZlBx0nVbX1ZMO0A+PDVbXTJNP6x9mKK8lK4M+q6gtt2j8CJo11tiW5Fngg8DjgJVX1J3Mdg9Svqp428TnJYXTHyuN7+r94PuLSwjDX23+q9fOgfXXI9GYt/iQF7FpVV7dpfwV4yGxNf4oxbA1cUVU7J3kh8Miq+pu5jGGBeCPwjqp6+3wHMqGqHjHxOcnrgQdV1Z/29H/aoPG0cS3YM/8by3x9I5+pxRr3KEl2Bm6sqv8B/gC4cJ5DWpDG6WrMOC2rNB820WPs0cC32udNoi2ZZpt/P+Cy2Y5Fgy3mvGzBJP9JHp3kwiS3JPk4sGVf/6cnuSjJTUm+luT3e/rtnuRbbdz/TPLxiVuGJi5BJXlVkp8AH5jC9O6b5BNJVif5QZKXjoj7nklOT/KzJOfTncXu7V9JHtQ+H5Dk8hbnj5P8bZKtgM8B922X6n7e5v/6JKcm+XCSnwGH9V4G7vHCJNcmuS7JK3rmu85tU72X4pJ8CNgF+H9tfq8ccNvDfdtyrUlydZI/75nW65OckuSDbVkuS7J8xOYdZjlwQc/ndSrsJE9J8p0kNyd5Z5IvJfmznv4vTHJFkrVJzkxyv771/uJ2CXRtkv9IksnGTee4JDe0+V6S5JGDgk9yjyTvb+v+x0nePNGwJnlXklN7hj02ydlt+hP75KuT3Jjulprn9Qx7Yhv/s0l+AfxRki2S/EuSHyW5Pt1l+7u04bdP8pm2L69J8pUkd2r9XtViu6Wty31b+Z2SHJXke0l+2rbndj0xHJLkh63f34/aiC3edyc5q83nS33b4qGt35oWw/8etawDpr9dkg+0/Xxtkk/19Pvztn+uafvrfaeyDyTZrK3PG5N8Hziwb57nJvmzJA8D3g08Lt2xclNP3G+eaRwaLCPq4CR3aet/bZLLgcf0jTu0PWj9h9b9fdNJhtQFvds/yUQ9OvH3u3Rn4Efu+wPmd/927NyS5Cxg+55+/fXzYUm+34b9QZLnTbKv9tcn691Wm+H10blZt969/dbWJF9uxRe3eT4n69/28bA2jZvStRXP6Ol3YjsezmjLcl6SddrQKRralvSsu0PT1Z83pqdOG3As98e/Msnfte3/i3R1/r3T3a5yS5IvJNm2L55h7fLQercnzsOT/Aj44qAFzZC6Jsn3gAdwR7u+xYBxd07yyXTH1U+TvKMnrtekq/NvSNe232OK62+y43Flkv2S7A+8GnhOi+/i1v/2/WuGcTw2yYp0udj1Sd42ZP1N1v6OamsH5pN9098syb+2af8gyUuy7rE7Km84LMlX2/zXtvF7r0IPHXeDVdW8/wF3Bn4I/DWwOfAs4LfAm1v/3YEbgD2AzYBDgZXAFj3jvqyN+yfAb3rG3Ru4FTi2DX+XSaZ3J7pK5LVt2g8Avg88dUjsJwOnAFsBjwR+DHy1p3/RXeYCuA54Qvu8LbB7T4yr+qb7+rYOntliuksr+3Drv6xN+2Nt3r8HrAb2a/1PnFgHg+bRlne/nu6J6S1p3V8C3kn3JWy3Nu19e2L7FXBAW39vBb7RM613Au8csb1fB9zUpvHL9vk24Ob2eTO6hu9nbXsuadv3t3SXtGnr5WrgYa3/a4Cv9a33zwDb0H3RWQ3sP9m4wFPb9t8GSBtmhyHL8SngPW393ws4H3hR63dX4LvAYcATgBuBnfr2ybfR7XNPAn4BPKRn290M7NW2/ZbAvwGnA9sBdwP+H/DWNvxb6Rr9zdvfE1rsDwGuAe7bs40f2D6/HPgGsFOL4T3Ax1q/hwM/B57Y+r2txbvfkPVwInBLz/Bvpx0Dbd1cA7ygrevd27p4xLBlHTD9M4CP0x0zmwNPauX7tGnt3ub7f4EvT3EfeDFwJbBzW6fnsO7+fy537GuH0XNM9x9fM4nDv4H708g6GDgG+ErbbjsDl9LqNiZvD4bW/QPiGFoX0Fe/9oyzP3Bti2vkvj9g3K9zR53wRLpjqr++X9Km+zPuqC924I7jadi+2l+f9O6/ezO6Prr9WBg0D3rauJ7pTWyPzenq2le3bbNPW67eum4N8Ni2bB8BTu6Z1meAo0bsK++nazN+Q1dnTbQlNwGX9a2799K1o48Cfg08bNC2ZHBb+Q3g3sCOdPvPhXRXG7agS9Jf1zevYe3yqHp3YtwPtnHvMmB5J6trVjK8nt4MuBg4rk1/S+Dxrd8L23Z6AN3tZZ8EPjTF9Tf0eOyPiZ4cpqf/udxR184kjq8Dh7TPWwN7DlkPezN6fx/V1k6Me3s+OWD6LwYub9t4W+ALrNu2jMobDqPLc/68ba+/oKtPMtm4G1zPzlWFPjKIrqK7fQFb2de4o3J6F/CmvnG+0zbaE+kS7t5xv8q6Fdtv6EkqJpneHsCP+vodDXxgyMH0W+ChPWX/yPDk/0fAi4C7D9gZByX/Xx5Q1t8Y9M77n4D3t88nMs3kn+4Avg24W0//twIn9sTxhZ5+Dwf+ZwO3+RLgCuA+wB8CZ/T1fz7w9Z7u0DWkE5XE54DDe/rfie6LxP161vvje/qfQmtERo1LV7l+F9gTuNOI+O9NV/HcpafsucA5Pd2PpWvYfgg8t29b3Aps1RffP/Rsuw/2LfsvaIl7K3sc8IP2+Y3Ap+lpgFv5g+gaqv2Azfv6XUH7Mte6d6Dbl5fQJV29DfBWdMfQqOS/d/it2/6zM/Ac4Ct9w7+HOxrLdZZ1wLR3AH4HbDug3/uBf+qb72+BZVPYB74IvLin31OYfvI/7Tj8G7jNR9bBdF8E9u/pdwR3JJuTtQdD6/4BcQytCxiQ/AMPbsfbxAmekft+X/kurF8nfJThyf9NwP+iL/kYsa9+cFj8TF4f3X4sDJoHo5P/JwA/6V1/dInx63vieF9PvwOAKzdwf9m2bactgf8D/Edf/4l1t1NP2fnAwYO2JYPbyuf1dH8CeFdP918Bn+qb17B2eVS9OzHuA0Ys62R1zUqG19OPo/sismRAv7OBv+zpfsiAuIatv6HHY39MTJ78zySOLwNvALafZH/ZmyH7O5O3tXvTl08OmP4X6UnI6drfiWN3ZN5Ad2xd3dPvrm3c+0w27ob+LZTbfu4L/Lja0jQ/7Pl8P+AV6S4b3pTucubObbxB417TN/3VVfWrKU7vfnS34PT2ezXdiu+3lG6D9s7vhwOGm/C/6Cq3H6a7vPu4EcMOWo7Jhvkh3TLM1H2BNVV1S9+0d+zp/knP518CW2YK978l2a2t07V0yemVdGdd927re+KB3/vSs2xt+/a+keJ+wNt7ttEaugN3VIxbTzZuVX0ReAfwH8D1SY5PcvcBi3I/urNa1/VM5z1038YnYj6frmIMXeXSa21V/aKnu3/b9W7XpXSVwAU98/qvVg7wz3RnSz6f7laAo9r8r6Y70/R64IYkJ+eO21HuB5zWM70r6BL2e7P+uv8F8NMB66BX7/A/p1unE8fTHn3H0/PoKrNBy9pvZ7p9ce2Afvel53hr8/0pU9sH1llGRh+3k5lJHFrfZHXwqG03WXswqu5fxwbUBbRbEz5NlzB/pWdek+37vXEPqhPW04Z5Dt0ZxuvS3TLz0EHD9pisLZmsPpqu+wLXVNXv+qY942MjyTPaOl1Ft65/ApwEPL+t7/5bUWdyDF7f8/l/BnT3T2tYuzyq3h00br+p1DXD7Az8sKpunWy67fNEsjphzuvSDYzjcLov4Fcm+WaSp4+Yz7D9fbK2FtbPJwctQ+/66K9/RuYNvctXVb9sH7ee4rhTtlCS/+uAHZN17oPdpefzNcBbqmqbnr+7VtXHhoy7c9/0q6971PSuofuW19vvblV1wIC4V9N9g+yd3y4DhuuCqPpmVR1Et7E+xR0JYX98w+IepH/e17bPv6DbiSf0Nzijpn0tsF2Su/VN+8dTiGekqrqoqrYB3gK8tn2+HHhUW9efbINeR3fZDOjuv+3tpttOL+rbTnepqq9NIYyR41bVv1fVHwCPoKtM/m7INH5Nd5ZhYhp3r3XfbHAk3aXBa4FX9o2/bbrnPSb0bjtYd/vcSNfAPKJnXveo9iapqrqlql5RVQ8A/j/gb9Lu7a+qj1b35o/7tWke2xP/0/rWwZZV9WO6dX/7fpXkrsA9R67RdYffmu6S6bVtPl/qm8/WVfUXQ5a13zV0++I2A/pd25ZrYr5btTinsp+us4yMOG4niW+mcWh9k9XBo7bdZO3BqLp/PVOpC9I9X/NRujNw7+mb12T7fm/cg+qEgarqzKp6Mt2Z4yvpboWA6bclo+qjydqSUa4Fdm7rqHfas9GWnN7ajw8Bh7XPa4ClbV2vmOKkZrJ8wwxrl0fVuxMma5unW9dcA+wy5CTdOtPljitR1w8Ytt9Gq0s3JI6quqqqnkuXXx0LnNq3T/catr+PbGunuAzr5C6sX/+MzBtGmMm461koyf/X6TbwS5MsaWd/H9vT/73Ai5Pskc5WSQ5syenX6b45v6SNe1DfuIOMmt75wM/SPdBxl3QPbzwyyWP6J1JVt9Hdk/b6JHdN8nC6e0jXk+TO6R7KukdV/Zbuns3bWu/rgXu2s0cb6h/avB9Bd2/px1v5RcAB6R6WvA/dGeBe19PdV7eeqrqG7rartybZMt0DcYfT3Y85W/4AuDDJnenOWnyvr/8ZwO8leWarrI5k3Ur53cDRbbknHoR59hTnPXTcJI9p+8XmdI3Cr7hjO92uqq4DPg/8a5K7p3tQ6YFJntSm82DgzcCfAocAr0yyW99k3tD2iycATwf+c1Cw7azZe4HjktyrTX/HJE9tn5+e5EEt4ZnYr25L8pAk+6R78OtXdJXaxLK8G3hL7njQeWk7dgBOBZ6e5PFt+7yRyeuKA3qGfxNwXtuPPgM8ON0DxJu3v8ekezhxUm09fw54Z5Jt2/hPbL0/Crwg3dWkLehuuTuvqlZOYdKn0NU3O6V7WO+oEcNeD+zUlm2QmcSh9U1WB59Cd/xum2QnutsuJkzWHoyq+9cx1bqA7kTGVnTPGfSa8r5fVT8EVnBHnfB4ui/y60n3sOkzWvLya7p73XvbklH76ijD6qOLgD9p7cyD6NqCXkPbEuA8unX3yrb8e7flOnka8Q0z0ZbcH7hukrOyg0zWVk7HsHZ5VL07FTOpa86nS0yPafv9lkn2av0+Bvx1uofOt27T/fiQqwT9Rh2P/a4HlmXdL4O9ph1Hkj9NsrS1lze14kHH64T19vfJ2topOgV4WRtvG+BVEz0myxtGmcm4gyyI5L+qfkP3YNZhdLeDPIcuqZ7ov4LuAYh3tP5Xt2F7xz2cboP/KV2l++sR8xs1vdvoKqfdgB/QfRN8HzAsMX8J3SWZn9DdO/iBEYt6CLAy3dt7XtxipaqupNvpv5/ucs6GXG79Uov/bOBfqurzrfxDdA/3rKTbYT7eN95bgde0+f3tgOk+l+4eu2uB0+juUz1rKgGlezr+3ZMMNvE6tt8Dvl21zmV6qupG4Nl090v+lO65ghW07VpVp9F9uz+5rc9LgSm9L3iSce9Od/CvpbsU+FPgX4ZM6vl0D7Fd3oY/Fdgh3ZeVDwPHVtXFVXUV3W0LH8odb2D4SRvnWrovVS9u+8Ewr6Lbzt9oMX+BO96lvWvr/jld8vPOqjqX7qrDMXT78E/ozoi8uo3zdrqHmj6f5Ba6h9D2aOvnMrovWx+layzWsu4tV4N8lO5B7jV02/Z5bVq30N1Pf3Bb1p9wx8NSU3UI3X2fV9LdU/3yNu2z6e7T/ESL84FtPlPxXuBMumPkQnrqmwG+SPf6vJ8kubG/5wzjUJ8p1MFvoDs2f0BXt32oZ9yR7cGoun+AqdYFz6V7LmBt7njjz/Omse//H7pjcA3dsfTBIcPdCXhFm+YaumfV/rL1G7mvjjCqPjqO7j7n6+luq+k/CfR64KTWlqzzNqO2PZ5BV7/eSPcyiOdPUtfdLt0bdV49ov/mdO3Ud+kegr1g2LAjTNZWTsewdnlovTsVM6lreo6rB9E9f7iKLtcCOIFuPXyZ7rj6FaOT+F5Dj8cBJr5Q/jTJoNexziSO/YHLkvycbj0fPOKL4Kj9fVRbOxXvpVsPl9C9fvazdCe3J76IDMwbpjjtmYy7jvTlXJuEJOcB766qUYm4FpF2pmAV3YNX58x3PDORKf6o22KR5ES6B7xeM9+xSP1sDyRNmOv2N92rOt9dVfebdOA5tCDO/M9UkicluU+7zHso8Pt0D2loEUvy1CTbtLPlr6Z7cPYb8xyWpAXM9kDSfGm3Kh7Q6p8d6a7inTbfcfVbtL9O1uchdPdZbU137/iz2v1RWtweR3c7ycRlrmdW92vAkjSM7YGk+RK6W6E+Tvec3Rl0r89eUDbJ234kSZIkrW+TuO1HkiRJ0uTm9Laf7bffvpYtWzaXs5SkReWCCy64saqWTj7kps32QpKGm0lbMafJ/7Jly1ixYqq/uyFJ4yfJTH4hc5NheyFJw82krfC2H0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmDD5lyRJksaEyb8kSZI0Jkz+JUmSpDFh8i9JkiSNCZN/SZIkaUzM6S/8SloYlh11xoynsfKYA2chEmn2zcb+PR0eE5IWA8/8S5IkSWPC5F+SJEkaEyb/kiRJ0pgw+ZckSZLGhMm/JEmSNCZM/iVJkqQxYfIvSZIkjQmTf0mSJGlMTJr8J9kyyflJLk5yWZI3tPL7JzkvyVVJPp7kzhs/XEmSJEnTNZUz/78G9qmqRwG7Afsn2RM4FjiuqnYF1gKHb7wwJUmSJM3UpMl/dX7eOjdvfwXsA5zayk8CnrlRIpQkSZI0K6Z0z3+SzZJcBNwAnAV8D7ipqm5tg6wCdtw4IUqSJEmaDVNK/qvqtqraDdgJeCzwsEGDDRo3yRFJViRZsXr16ulHKkmSJGlGNuhtP1V1E3AusCewTZIlrddOwLVDxjm+qpZX1fKlS5fOJFZJkiRJMzCVt/0sTbJN+3wXYD/gCuAc4FltsEOBT2+sICVJkiTN3JLJB2EH4KQkm9F9WTilqj6T5HLg5CRvBr4FvH8jxilJkiRphiZN/qvqEuDRA8q/T3f/vyRpE9Su+r4PeCTdc10vBL4DfBxYBqwE/ndVrU0S4O3AAcAvgcOq6sI2nUOB17TJvrmqTprDxZAk9fAXfiVJw7wd+K+qeijwKLpbPo8Czm6/8XJ26wZ4GrBr+zsCeBdAku2A1wF70J0wel2SbedyISRJdzD5lyStJ8ndgSfSbumsqt+0lz4cRPfbLrDub7wcBHyw/TbMN+heCrED8FTgrKpaU1Vr6V4Xvf8cLookqYfJvyRpkAcAq4EPJPlWkvcl2Qq4d1VdB9D+36sNvyNwTc/4E7//Mqx8Pb4aWpI2PpN/SdIgS4DdgXdV1aOBX3DHLT6DZEBZjShfv9BXQ0vSRmfyL0kaZBWwqqrOa92n0n0ZuL7dzkP7f0PP8Dv3jD/x+y/DyiVJ88DkX5K0nqr6CXBNkoe0on2By4HT6X7bBdb9jZfTgeensydwc7st6EzgKUm2bQ/6PqWVSZLmwVTe8y9JGk9/BXwkyZ2B7wMvoP3eS5LDgR8Bz27DfpbuNZ9X073q8wUAVbUmyZuAb7bh3lhVa+ZuESRJvUz+JUkDVdVFwPIBvfYdMGwBRw6ZzgnACbMbnSRpOrztR5IkSRoTJv+SJEnSmDD5lyRJksaEyb8kSZI0Jkz+JUmSpDFh8i9JkiSNCV/1KWlalh11xqxMZ+UxB87KdCRJ0uQ88y9JkiSNCZN/SZIkaUyY/EuSJEljwuRfkiRJGhMm/5IkSdKYMPmXJEmSxoSv+pQWkdl6vaYkSRpPnvmXJEmSxoTJvyRJkjQmJk3+k+yc5JwkVyS5LMnLWvnrk/w4yUXt74CNH64kSZKk6ZrKPf+3Aq+oqguT3A24IMlZrd9xVfUvGy88SZIkSbNl0uS/qq4Drmufb0lyBbDjxg5MkiRJ0uzaoLf9JFkGPBo4D9gLeEmS5wMr6K4OrB0wzhHAEQC77LLLDMOVtKmZjTcYrTzmwFmIRJKkTd+UH/hNsjXwCeDlVfUz4F3AA4Hd6K4M/Oug8arq+KpaXlXLly5dOgshS5IkSZqOKSX/STanS/w/UlWfBKiq66vqtqr6HfBe4LEbL0xJkiRJMzWVt/0EeD9wRVW9rad8h57B/hi4dPbDkyTNlyQrk3y7vdFtRSvbLslZSa5q/7dt5Uny70muTnJJkt17pnNoG/6qJIfO1/JIkqZ2z/9ewCHAt5Nc1MpeDTw3yW5AASuBF22UCCVJ8+mPqurGnu6jgLOr6pgkR7XuVwFPA3Ztf3vQ3Rq6R5LtgNcBy+naiwuSnD7oGTFJ0sY3lbf9fBXIgF6fnf1wJEkL3EHA3u3zScC5dMn/QcAHq6qAbyTZpl0h3hs4q6rWALRXRe8PfGxuw5Ykgb/wK0karoDPJ7mgvbkN4N7tFdATr4K+VyvfEbimZ9xVrWxY+XqSHJFkRZIVq1evnsXFkCRN2KBXfUqSxspeVXVtknsBZyW5csSwg64Q14jy9QurjgeOB1i+fPnAYSRJM+OZf0nSQFV1bft/A3Aa3Vvdrp944UP7f0MbfBWwc8/oOwHXjiiXJM0Dk39J0nqSbJXkbhOfgafQvdXtdGDijT2HAp9un08Hnt/e+rMncHO7LehM4ClJtm1vBnpKK5MkzQNv+5EkDXJv4LTubc8sAT5aVf+V5JvAKUkOB34EPLsN/1ngAOBq4JfACwCqak2SNwHfbMO9ceLhX0nS3DP5lyStp6q+DzxqQPlPgX0HlBdw5JBpnQCcMNsxSpI2nLf9SJIkSWPC5F+SJEkaEyb/kiRJ0pgw+ZckSZLGhMm/JEmSNCZM/iVJkqQxYfIvSZIkjQmTf0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmDD5lyRJksaEyb8kSZI0Jkz+JUmSpDFh8i9JkiSNCZN/SZIkaUyY/EuSJEljYtLkP8nOSc5JckWSy5K8rJVvl+SsJFe1/9tu/HAlSZIkTddUzvzfCryiqh4G7AkcmeThwFHA2VW1K3B265YkSZK0QE2a/FfVdVV1Yft8C3AFsCNwEHBSG+wk4JkbK0hJkiRJM7dB9/wnWQY8GjgPuHdVXQfdFwTgXkPGOSLJiiQrVq9ePbNoJUmSJE3blJP/JFsDnwBeXlU/m+p4VXV8VS2vquVLly6dToySJEmSZsGUkv8km9Ml/h+pqk+24uuT7ND67wDcsHFClCTNlySbJflWks+07vsnOa+97OHjSe7cyrdo3Ve3/st6pnF0K/9OkqfOz5JIkmBqb/sJ8H7giqp6W0+v04FD2+dDgU/PfniSpHn2MrpnvSYcCxzXXvawFji8lR8OrK2qBwHHteFoL4g4GHgEsD/wziSbzVHskqQ+UznzvxdwCLBPkova3wHAMcCTk1wFPLl1S5I2EUl2Ag4E3te6A+wDnNoG6X3ZQ+9LIE4F9m3DHwScXFW/rqofAFcDj52bJZAk9Vsy2QBV9VUgQ3rvO7vhSJIWkH8DXgncrXXfE7ipqm5t3avo3v5G+38NQFXdmuTmNvyOwDd6ptk7zjqSHAEcAbDLLrvM3lJIkm7nL/xKktaT5OnADVV1QW/xgEFrkn6jxlm30BdESNJGN+mZf0kzt+yoM+Y7BGlD7QU8o93muSVwd7orAdskWdLO/u8EXNuGXwXsDKxKsgS4B7Cmp3xC7ziSpDnmmX9J0nqq6uiq2qmqltE9sPvFqnoecA7wrDZY78seel8C8aw2fLXyg9vbgO4P7AqcP0eLIUnq45l/SdKGeBVwcpI3A9+iexsc7f+HklxNd8b/YICquizJKcDlwK3AkVV129yHLUkCk39J0iSq6lzg3Pb5+wx4W09V/Qp49pDx3wK8ZeNFKEmaKm/7kSRJksaEyb8kSZI0Jkz+JUmSpDFh8i9JkiSNCZN/SZIkaUyY/EuSJEljwuRfkiRJGhMm/5IkSdKYMPmXJEmSxoS/8Ctp0Vt21BmzMp2Vxxw4K9ORJGmh8sy/JEmSNCZM/iVJkqQxYfIvSZIkjQmTf0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmDD5lyRJksaEyb8kSZI0JiZN/pOckOSGJJf2lL0+yY+TXNT+Dti4YUqSJEmaqamc+T8R2H9A+XFVtVv7++zshiVJkiRptk2a/FfVl4E1cxCLJEmSpI1oyQzGfUmS5wMrgFdU1dpBAyU5AjgCYJdddpnB7KQNs+yoM2ZlOiuPOXBWpiMtJkm2BL4MbEHXVpxaVa9Lcn/gZGA74ELgkKr6TZItgA8CfwD8FHhOVa1s0zoaOBy4DXhpVZ0518sjSepM94HfdwEPBHYDrgP+ddiAVXV8VS2vquVLly6d5uwkSXPs18A+VfUourp+/yR7AsfS3fa5K7CWLqmn/V9bVQ8CjmvDkeThwMHAI+huIX1nks3mdEkkSbebVvJfVddX1W1V9TvgvcBjZzcsSdJ8qs7PW+fm7a+AfYBTW/lJwDPb54NaN63/vknSyk+uql9X1Q+Aq7HNkKR5M63kP8kOPZ1/DFw6bFhJ0uKUZLMkFwE3AGcB3wNuqpxTitcAABRaSURBVKpb2yCrgB3b5x2BawBa/5uBe/aWDxinf35HJFmRZMXq1atne3EkSUzhnv8kHwP2BrZPsgp4HbB3kt3ozgKtBF60EWOUJM2DqroN2C3JNsBpwMMGDdb+Z0i/YeWD5nc8cDzA8uXLBw4jSZqZSZP/qnrugOL3b4RYJEkLUFXdlORcYE9gmyRL2tn9nYBr22CrgJ2BVUmWAPege1PcRPmE3nEkSXPMX/iVJK0nydJ2xp8kdwH2A64AzgGe1QY7FPh0+3x666b1/2JVVSs/OMkW7U1BuwLnz81SSJL6zeRVn5KkTdcOwEntzTx3Ak6pqs8kuRw4OcmbgW9xx5Xg9wMfSnI13Rn/gwGq6rIkpwCXA7cCR7bbiSRJ88DkX5K0nqq6BHj0gPLvM+BtPVX1K+DZQ6b1FuAtsx2jJGnDeduPJEmSNCZM/iVJkqQxYfIvSZIkjQmTf0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmDD5lyRJksaEyb8kSZI0Jkz+JUmSpDGxZL4DGGfLjjpjxtNYecyBsxCJJEmSxoFn/iVJkqQxYfIvSZIkjQmTf0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmFg0r/qcjddigq/G1IabrX1PkiRpvnnmX5IkSRoTJv+SJEnSmJg0+U9yQpIbklzaU7ZdkrOSXNX+b7txw5QkSZI0U1M5838isH9f2VHA2VW1K3B265YkSZK0gE2a/FfVl4E1fcUHASe1zycBz5zluCRJ8yjJzknOSXJFksuSvKyVD7zym86/J7k6ySVJdu+Z1qFt+KuSHDpfyyRJmv49//euqusA2v97DRswyRFJViRZsXr16mnOTpI0x24FXlFVDwP2BI5M8nCGX/l9GrBr+zsCeBd0XxaA1wF7AI8FXuetopI0fzb6A79VdXxVLa+q5UuXLt3Ys5MkzYKquq6qLmyfbwGuAHZk+JXfg4APVucbwDZJdgCeCpxVVWuqai1wFuvfSipJmiPTTf6vb5U67f8NsxeSJGkhSbIMeDRwHsOv/O4IXNMz2qpWNqxckjQPppv8nw5M3Ld5KPDp2QlHkrSQJNka+ATw8qr62ahBB5TViPJB8/I2UUnayKbyqs+PAV8HHpJkVZLDgWOAJye5Cnhy65YkbUKSbE6X+H+kqj7Ziodd+V0F7Nwz+k7AtSPK1+NtopK08S2ZbICqeu6QXvvOciySpAUiSYD3A1dU1dt6ek1c+T2Gda/8ng68JMnJdA/33lxV1yU5E/jHnod8nwIcPRfLIEla36TJvyRpLO0FHAJ8O8lFrezVdEn/Ke0q8I+AZ7d+nwUOAK4Gfgm8AKCq1iR5E/DNNtwbq6r/9dGSpDli8j8Ny446Y75DkKSNqqq+yuD79WHAld+qKuDIIdM6AThh9qKTJE3XRn/VpyRJkqSFweRfkiRJGhMm/5IkSdKYMPmXJEmSxoTJvyRJkjQmxu5tP76pZ+OajfW78pgDZyESSZIk9fPMvyRJkjQmTP4lSZKkMWHyL0mSJI0Jk39JkiRpTJj8S5IkSWPC5F+SJEkaEyb/kiRJ0pgw+ZckSZLGhMm/JEmSNCZM/iVJkqQxYfIvSZIkjQmTf0mSJGlMLJnvAKR+y446Y75DkCRJ2iR55l+SJEkaEyb/kiRJ0piY0W0/SVYCtwC3AbdW1fLZCEqSJEnS7JuNe/7/qKpunIXpSJIkSdqIvO1HkjRQkhOS3JDk0p6y7ZKcleSq9n/bVp4k/57k6iSXJNm9Z5xD2/BXJTl0PpZFktSZ6Zn/Aj6fpID3VNXx/QMkOQI4AmCXXXaZ4ezUzzfjSLNnto6nlcccOCvTWQBOBN4BfLCn7Cjg7Ko6JslRrftVwNOAXdvfHsC7gD2SbAe8DlhO12ZckOT0qlo7Z0shSbrdTM/871VVu9NV+kcmeWL/AFV1fFUtr6rlS5cuneHsJElzpaq+DKzpKz4IOKl9Pgl4Zk/5B6vzDWCbJDsATwXOqqo1LeE/C9h/40cvSRpkRsl/VV3b/t8AnAY8djaCkiQtWPeuqusA2v97tfIdgWt6hlvVyoaVryfJEUlWJFmxevXqWQ9ckjSD5D/JVknuNvEZeApw6eixJEmbqAwoqxHl6xd6pViSNrqZnPm/N/DVJBcD5wNnVNV/zU5YkqQF6vp2Ow/t/w2tfBWwc89wOwHXjiiXJM2DaT/wW1XfBx41i7FIkha+04FDgWPa/0/3lL8kycl0D/zeXFXXJTkT+MeJtwLRXSU+eo5jnhPz9QKGTegBc0lzYDbe8y9J2gQl+RiwN7B9klV0b+05BjglyeHAj4Bnt8E/CxwAXA38EngBQFWtSfIm4JttuDdWVf9DxJKkOWLyL0kaqKqeO6TXvgOGLeDIIdM5AThhFkOTJE2TP/IlSZIkjQmTf0mSJGlMmPxLkiRJY8LkX5IkSRoTJv+SJEnSmDD5lyRJksaEyb8kSZI0Jkz+JUmSpDFh8i9JkiSNCZN/SZIkaUyY/EuSJEljwuRfkiRJGhMm/5IkSdKYMPmXJEmSxoTJvyRJkjQmTP4lSZKkMWHyL0mSJI2JJfMdgCRJmr5lR50xL/NdecyB8zJfSTPjmX9JkiRpTJj8S5IkSWPC5F+SJEkaEyb/kiRJ0piY0QO/SfYH3g5sBryvqo6ZlagkSZsU24tNjw8aS4vTtJP/JJsB/wE8GVgFfDPJ6VV1+WwFJ0la/GwvNJvm40uHXzi0KZnJmf/HAldX1fcBkpwMHARYmUuSetleaFGbr6sc82W+vux4NWluzCT53xG4pqd7FbBH/0BJjgCOaJ0/T/Kdacxre+DGaYy3kLlMi4PLtDgsqGXKsTMa/X6zFMZCMlftxYLaD/os5NhgYcdnbNM3rfhmWIdN1YJZd0OWd8HEN8D2zKCtmEnynwFltV5B1fHA8TOYD0lWVNXymUxjoXGZFgeXaXHYFJdpEzMn7cVC3g8WcmywsOMztulbyPEt5NhgYcfXYls23fFn8rafVcDOPd07AdfOYHqSpE2T7YUkLRAzSf6/Ceya5P5J7gwcDJw+O2FJkjYhtheStEBM+7afqro1yUuAM+le3XZCVV02a5Gta0a3DS1QLtPi4DItDpviMm0y5rC9WMj7wUKODRZ2fMY2fQs5voUcGyzs+GZ2O33VerddSpIkSdoE+Qu/kiRJ0pgw+ZckSZLGxIJP/pPsn+Q7Sa5OctR8xzNVSU5IckOSS3vKtktyVpKr2v9tW3mS/HtbxkuS7D5/kQ+XZOck5yS5IsllSV7WyhftciXZMsn5SS5uy/SGVn7/JOe1Zfp4e0iRJFu07qtb/2XzGf8wSTZL8q0kn2ndi3p5AJKsTPLtJBclWdHKFu2+p9mz0NqJDa0r5ynGKdUR8xDXNklOTXJlW3+PW2Dr7a/bNr00ycdaGzJv624h5xpDYvvntm0vSXJakm16+h3dYvtOkqfOdWw9/f42SSXZvnXPeXsyLL4kf9XWz2VJ/qmnfIPW3YJO/nPHT8I/DXg48NwkD5/fqKbsRGD/vrKjgLOralfg7NYN3fLt2v6OAN41RzFuqFuBV1TVw4A9gSPb9ljMy/VrYJ+qehSwG7B/kj2BY4Hj2jKtBQ5vwx8OrK2qBwHHteEWopcBV/R0L/blmfBHVbVbz7uXF/O+p1mwQNuJDa0r58NU64i59nbgv6rqocCj6GJcEOstyY7AS4HlVfVIuofXD2Z+192JLNxcY1BsZwGPrKrfB74LHA3Qjo+DgUe0cd7Zju25jI0kOwNPBn7UUzwf7cl68SX5I7pfRv/9qnoE8C+tfMPXXVUt2D/gccCZPd1HA0fPd1wbEP8y4NKe7u8AO7TPOwDfaZ/fAzx30HAL+Q/4NN1BskksF3BX4EK6Xx69EVjSym/fD+neVvK49nlJGy7zHXvfcuxEV+HvA3yG7geWFu3y9CzXSmD7vrJNYt/zb0b7xYJvJyarK+chninXEXMc192BH/TXQQtovU38UvV2rb78DPDU+V53CznX6I+tr98fAx9pn9c5bnvbprmMDTiV7kvn7e3NfLUnA7brKcB+A4bb4HW3oM/8M/gn4Xecp1hmw72r6jqA9v9erXzRLWe620MeDZzHIl+udvn7IuAGurMS3wNuqqpb2yC9cd++TK3/zcA95zbiSf0b8Ergd637nizu5ZlQwOeTXJDkiFa2qPc9zYoFva2nWFfOtQ2pI+bSA4DVwAfaLUnvS7IVC2S9VdWP6c62/gi4jq6+vICFse56LZZ68YXA59rneY8tyTOAH1fVxX295j225sHAE9otZl9K8phWvsHxLfTkf0o/Cb8JWFTLmWRr4BPAy6vqZ6MGHVC24Jarqm6rqt3ozoY9FnjYoMHa/wW9TEmeDtxQVRf0Fg8YdFEsT5+9qmp3ukuwRyZ54ohhF9NyaWYW7LbegLpyzkyjjphLS4DdgXdV1aOBXzC/t0ato907fxBwf+C+wFZ09VG/BbH/DbBQtjNJ/p7u9riPTBQNGGzOYktyV+DvgdcO6j2gbL6Oj23pbiX8O+CUJGEa8S305H9T+0n465PsAND+39DKF81yJtmcrjH7SFV9shUv+uUCqKqbgHPpDqxtkkz8CF5v3LcvU+t/D2DN3EY60l7AM5KsBE6mu6z/byze5bldVV3b/t8AnEb3RW2T2Pc0IwtyW29gXTmXNrSOmEurgFVVdV7rPpXuy8BCWG8A+wE/qKrVVfVb4JPAH7Iw1l2vBV0vJjkUeDrwvGr3qSyA2B5I96Xu4nZs7ARcmOQ+CyC2CauAT1bnfLord9tPJ76Fnvxvaj8JfzpwaPt8KN19oBPlz29PlO8J3DxxyW4had8w3w9cUVVv6+m1aJcrydKJtw0kuQtd5X4FcA7wrDZY/zJNLOuzgC/2VF7zrqqOrqqdqmoZ3fHyxap6Hot0eSYk2SrJ3SY+A08BLmUR73uaNQuunZhGXTlnplFHzGVsPwGuSfKQVrQvcDkLYL01PwL2THLXto0n4pv3dddnwdaLSfYHXgU8o6p+2dPrdODgdG+guz/dw7Xnz1VcVfXtqrpXVS1rx8YqYPe2T877ems+RfdlnSQPBu5M97zJhq+7jf3Awiw88HAA3RPh3wP+fr7j2YC4P0Z3T+Bv6Xaiw+nuqzwbuKr9364NG7q3VXwP+DbdmwTmfRkGLNPj6S4lXQJc1P4OWMzLBfw+8K22TJcCr23lD2gHz9XAfwJbtPItW/fVrf8D5nsZRizb3sBnNoXlafFf3P4um6gLFvO+59+s7h8Lqp3Y0LpyHuOctI6Yh5h2A1a0dfcputscFsx6A94AXNnaiw8BW8znumMB5xpDYrua7v70iePi3T3D/32L7TvA0+Y6tr7+K7njgd85b0+GrLs7Ax9u+96FdG8qnNa6SxtJkiRJ0iZuod/2I0mSJGmWmPxLkiRJY8LkX5IkSRoTJv+SJEnSmDD5lyRJksaEyb8WtSRvTbJ3kmcmWTC/BClJ2vjmqg1o8/hM+/yMiXm1+T68Z7g3JtlvY8UhzQaTfy12ewDnAU8CvjLPsUiS5tactwFVdXpVHdM6nwk8vKffa6vqC3MRhzRdJv9alJL8c5JLgMcAXwf+DHhXktcmeUySS5J8vQ13aRtns9b9zdb/Ra187yTnJjk1yZVJPtJ+vZEkf5DkS0kuSHJmz0+mvzTJ5W06J8/PWpCk8TRJG/CgJF9IcnGSC5M8sP066z8nuTTJt5M8p01nVP2/fyv7KvAnPfM+LMk7kvwh8Azgn5Nc1OZzYpJnteH2TfKtNr8TkmzRylcmeUOL7dtJHtrKn9Smc1Eb725zuEo1RpbMdwDSdFTV3yX5T+AQ4G+Ac6tqL4CW7B9RVV9LckzPaIfT/Sz3Y1ol/N9JPt/6PRp4BHAt8N/AXknOA/4vcFBVrW6NxVuAFwJHAfevql8n2WbjL7EkacIkbcB5wDFVdVqSLelOdP4J3a8HPwrYHvhmki+3yQ2q/1cA7wX2oftV2o8PiOFrSU6n+5XkU9u8af+3BE4E9q2q7yb5IPAXwL+10W+sqt2T/CXwt3RfXv4WOLKq/jvJ1sCvZmdtSevyzL8Ws0fT/Tz4Q4HLAVoifreq+lob5qM9wz8FeH6Si+guE98T2LX1O7+qVlXV79o0lwEPAR4JnNXGeQ2wUxv+EuAjSf4UuHXjLJ4kaYRBbcDdgB2r6jSAqvpVVf0SeDzwsaq6raquB75Ed9UABtf/DwV+UFVXVVUBH97A2B7Sxv9u6z4JeGJP/0+2/xe0+UH3xeNtSV4KbFNVti3aKDzzr0UnyW50Z1R2Am4E7toV5yLgaaNGBf6qqs7sm97ewK97im6jOzYCXFZVjxswrQPpKvJnAP+Q5BFW1JK08U3SBjxp2GgjJjmo/geomYQ5Sf+Jed4+v6o6JskZwAHAN5LsV1VXziAGaSDP/GvRqaqLqmo34Lt0D1p9EXhqVe1WVdcBtyTZsw1+cM+oZwJ/kWRzgCQPTrLViFl9B1ia5HFt+M2TPCLJnYCdq+oc4JXANsDWs7mMkqTBJmkDbgZWJXkmQJItktwV+DLwnPbs11K6kzfnj5jNlcD9kzywdT93yHC3AIPuzb8SWJbkQa37ELqrDUMleWBVfbuqjgVW0F19kGadyb8WpVZ5r22XaR9aVZf39D4cOD7J1+nOvtzcyt9Hd2n4wvZcwHsYcfWrqn4DPAs4NsnFdJeD/xDYDPhwkm8D3wKOq6qbZnUBJUlDTdIGHAK8tD0Q/DXgPsBpdLdrXkz3ZeGVVfWTYdOvql8BRwBntAd+fzhk0JOBv2sP6D6wb/wXAP/Z2orfAe+eZLFe3h5Ivhj4H+BzkwwvTUu6W9mkTUeSravq5+3zUcAOVfWyeQ5LkiRp3nnPvzZFByY5mm7//iFw2PyGI0mStDB45l+SJEkaE97zL0mSJI0Jk39JkiRpTJj8S5IkSWPC5F+SJEkaEyb/kiRJ0pj4/wEa8XVJHOpVxAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 936x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig,ax = plt.subplots(1,2,figsize=(13,4))\n",
    "ax[0].hist(degree_distr, bins=17)\n",
    "ax[0].set_title('degree distribution: #genes expressed per condition')\n",
    "ax[0].set_xlabel('#genes')\n",
    "\n",
    "ax[1].hist(edge_size_distr, bins=10);\n",
    "ax[1].set_title(f'edge size distribution: #number of conditions per gene')\n",
    "ax[1].set_xlabel('#conditions')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The shape, info, degree distribution, and edge size distribution are directly related to the same functions applied to the `H_trans_d` hypergraph. The shape will swap the vertex and edge count. The info will swap row and column counts, invert aspect ratio, and have the same nonzero cell count and density. The degree distribution of `H_trans` is the same as the edge size distribution of `H_trans_d` while the edge size distribution of `H_trans` is the same as the degree distribution of `H_trans_d`. Because of this direct relationship we will not show these properties of `H_trans_d`.\n",
    "\n",
    "Next we look at some properties that will not be directly related between `H_trans` and `H_trans_d`: pairwise intersection size distribution and s-diameters.\n",
    "\n",
    "### Edge intersection size distribution of `H_trans`\n",
    "The intersection of two hyperedges (representing genes, as in `H_trans`) identifies the conditions that both genes are significantly perturbed within. We write a function to compute the pairwise intersection size and then apply it to the `H_trans` hypergraph. Later we will apply it to `H_trans_d`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# BRENDA: need your help in writing this properly\n",
    "def pairwise_intersections(H, aggregated=True):\n",
    "    E = list(H.edges())\n",
    "    inters = []\n",
    "    for e in E:\n",
    "        ve = set(e.elements)\n",
    "        for f in E:\n",
    "            if e == f:\n",
    "                continue\n",
    "            vf = set(f.elements)\n",
    "            if len(vf.intersection(ve)) > 0:\n",
    "                inters += [len(vf.intersection(ve))]\n",
    "            \n",
    "    if aggregated:\n",
    "        vals, counts = np.unique(inters, return_counts=True)\n",
    "        inters_dist = dict(zip(vals, counts))\n",
    "        return inters_dist\n",
    "    else:\n",
    "        return inters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'H' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-24-97bb0f24d793>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mH\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mincidence_matrix\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdot\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mT\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'H' is not defined"
     ]
    }
   ],
   "source": [
    "m = H.incidence_matrix()\n",
    "a = m.dot(m.T)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see next that the pairwise edge intersection size distribution for `H_trans` is similar to the edge size distribution itself."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "# inters = pairwise_intersections(H_trans, aggregated=True)\n",
    "MW = H_trans.edge_adjacency_matrix(weighted=True)\n",
    "# inters = np.array([MW[i,j] for i in range(MW.shape[0]) for j in range(i+1,MW.shape[0])])\n",
    "# inters = np.array(np.unique(inters,return_counts=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "uMW = csr_matrix(triu(MW,1)).reshape(1,8030*8030).todense()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "IndexError",
     "evalue": "index 1 is out of bounds for axis 0 with size 1",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-23-13d351bfc996>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mj\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mn\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m         \u001b[0mcoords\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mctr\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0muMW\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mj\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m         \u001b[0mctr\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m//anaconda3/envs/hnx/lib/python3.8/site-packages/numpy/matrixlib/defmatrix.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, index)\u001b[0m\n\u001b[1;32m    191\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    192\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 193\u001b[0;31m             \u001b[0mout\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mN\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mndarray\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getitem__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    194\u001b[0m         \u001b[0;32mfinally\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mIndexError\u001b[0m: index 1 is out of bounds for axis 0 with size 1"
     ]
    }
   ],
   "source": [
    "n = 8030\n",
    "m = int(np.floor((n*(n-1))/2))\n",
    "coords = np.zeros(m)\n",
    "ctr = 0\n",
    "for i in range(n):\n",
    "    for j in range(i+1,n):\n",
    "        coords[ctr] = uMW[i,j]\n",
    "        ctr += 1\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inters = dict(inters)\n",
    "plt.scatter(inters.keys(), inters.values())\n",
    "plt.yscale('log')\n",
    "plt.xscale('log')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Edge diameters of `H_trans`\n",
    "The s-edge diameter of a hypergraph is the length of the longest s-distance within the hypergraph. For a disconnected hypergraph the `edge_diameters` function computes the s-edge diameter for each s-component and returns (1) the maximum diameter over all components, (2) the list of all diameters, and (3) the list of edge sets for each component. We compute the s=100-edge diameters for `H_trans`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D, diams, comps = H_trans.edge_diameters(s=100)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A component with only one edge has s-diameter 0 and is uninteresting, so we restrict to the non-singleton components. An element of `non_singleton_comps` is of the form (index in comps, diameter of component, number of edges in component).  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_singleton_comps = [(i, diams[i], len(comps[i])) for i in range(len(comps)) if len(comps[i]) > 1]\n",
    "len(non_singleton_comps)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since there is only one non-singleton s=100-component we see what its edges are and visualize the hypergraph."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "comps[non_singleton_comps[0][0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "hnx.draw(H_trans.restrict_to_edges(comps[non_singleton_comps[0][0]]).remove_static(), with_node_labels=False, with_edge_labels=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Edge intersection size distribution of `H_trans_d`\n",
    "Now we compute the edge intersection size distribution for the `H_trans_d` hypergraph. The intersection of two condition hyperedges counts the number of genes those two edges have in common. We display the distribution on a log-log plot again but can see that it does not have the typical power law shape. The edge sizes in `H_trans_d` are much larger than those in `H_trans` (around 500 in `H_trans_d` and only 100 for `H_trans`) and with a much different distribution (see vertex degree distribution for `H_trans`), so it makes sense that the intersection size distribution would be much different. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "inters_d = pairwise_intersections(H_trans_d, aggregated=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.scatter(inters_d.keys(), inters_d.values())\n",
    "plt.yscale('log')\n",
    "plt.xscale('log')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Edge diameters of `H_trans_d`\n",
    "We again compute the s=100-edge diameters and find the non-singleton components."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D_d, diams_d, comps_d = H_trans_d.edge_diameters(s=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_singleton_comps_d = [(i, diams_d[i], len(comps_d[i])) for i in range(len(comps_d)) if len(comps_d[i]) > 1]\n",
    "len(non_singleton_comps_d)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "There are four non-singleton s=100-components in `H_trans_d`. Next we visualize the smaller ones (there are 3 of them)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, axs = plt.subplots(1, 3, figsize=(15, 5))\n",
    "counter = 0\n",
    "for nsc in non_singleton_comps_d:\n",
    "    if len(comps_d[nsc[0]]) < 10:\n",
    "        hnx.draw(H_trans_d.restrict_to_edges(comps_d[nsc[0]]), with_node_labels=False, with_edge_labels=True, ax=axs[counter])\n",
    "        counter += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Applying s-Centrality measures\n",
    "In some related work we computed average s-betweenness centrality and average harmonic s-closeness centrality in `H_trans` for s between 1 and 50 to discover the most central hyperedges (genes) for this set of conditions. We found that immune response genes tended towards the higher centrality values. Here we show how to compute average s-centralities in general, but to avoid the high computation time we do not do the full computation."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def avg_s_betweenness_centrality(H, mins=1, maxs=1, normalized=True, verbose=False):\n",
    "    # First compute all s-betweenness centralities for s between mins and maxs (inclusive)\n",
    "    betc = {}\n",
    "    for s in range(mins, maxs+1, 1):\n",
    "        if verbose: # for a large range of s values, or for low s values, a status update can be helpful\n",
    "            print(s)\n",
    "        betc[s] = hnx.s_betweenness_centrality(H, s=s, normalized=normalized)\n",
    "        \n",
    "    # For each gene in the betc[mins] dictionary we compute its average centrality\n",
    "    # value across all s values. If that gene does not appear in a dictionary for \n",
    "    # a higher s value we say that it has centrality 0 for that s value.\n",
    "    gene_betc_list = {}\n",
    "    avg_betcs = {}\n",
    "    for gene in betc[mins]:\n",
    "        gene_betc_list[gene] = [betc[s][gene] if gene in betc[s] else 0 for s in betc]\n",
    "        avg_betcs[gene] = np.mean(gene_betc_list[gene])\n",
    "    \n",
    "    return betc, gene_betc_list, avg_betcs\n",
    "\n",
    "def avg_s_harmonic_closeness_centrality(H, mins=1, maxs=1, verbose=False):\n",
    "    # First compute all harmonic s-closeness centralities for s between mins and maxs (inclusive)\n",
    "    cloc = {}\n",
    "    for s in range(mins, maxs+1, 1):\n",
    "        if verbose: # for a large range of s values, or for low s values, a status update can be helpful\n",
    "            print(s)\n",
    "        cloc[s] = hnx.s_harmonic_closeness_centrality(H, s=s)\n",
    "        \n",
    "    # For each gene in the cloc[mins] dictionary we compute its average centrality\n",
    "    # value across all s values. If that gene does not appear in a dictionary for \n",
    "    # a higher s value we say that it has centrality 0 for that s value.\n",
    "    gene_cloc_list = {}\n",
    "    avg_clocs = {}\n",
    "    for gene in cloc[mins]:\n",
    "        gene_cloc_list[gene] = [cloc[s][gene] if gene in cloc[s] else 0 for s in cloc]\n",
    "        avg_clocs[gene] = np.mean(gene_cloc_list[gene])\n",
    "    \n",
    "    return cloc, gene_cloc_list, avg_clocs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "betc, gene_betc_list, avg_betcs = avg_s_betweenness_centrality(H_trans, mins=50, maxs=52)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cloc, gene_cloc_list, avg_clocs = avg_s_harmonic_closeness_centrality(H_trans, mins=50, maxs=52)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "{k: v for k, v in sorted(avg_betcs.items(), key=lambda item: item[1], reverse=True)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import hypernetx as hnx\n",
    "system = hnx.Hypergraph()\n",
    "edge0 = hnx.Entity('edge0')\n",
    "edge1 = hnx.Entity('edge1')\n",
    "node0 = hnx.Entity('node0')\n",
    "node1 = hnx.Entity('node1')\n",
    "\n",
    "system.add_edge(edge0)\n",
    "system.add_edge(edge1)\n",
    "system.add_node_to_edge(node0,edge0)\n",
    "system.add_node_to_edge(node0,edge1)\n",
    "system.add_node_to_edge(node1,edge1)\n",
    "\n",
    "#Memberships seen from the nodes stored within _edges\n",
    "print(\"Memberships seen from the nodes stored within _edges\")\n",
    "for el in system.edges:\n",
    "    for nd in system.edges[el]:\n",
    "        print(nd,system.edges[el][nd].memberships)\n",
    "        \n",
    "#Memberships seen from the nodes stored within _nodes\n",
    "print(\"Memberships seen from the nodes stored within _nodes\")\n",
    "for el in system.nodes:\n",
    "    print(el,system.nodes[el].memberships)\n",
    "\n",
    "hnx.drawing.rubber_band.draw(system)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "node0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "system.edges[edge1].elements[\"node0\"].memberships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "node0.memberships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "node1.memberships"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
